Support times better

JPMoresmau opened the pull request:
Add support for full precision of subseconds, so `PARAMETRIC_DATETIME` works properly.

Add type wrappers in the manner of `Numeric` for the various date/time data types, so we can avoid casts and not leak the accepted formats to the client.

nineinchnick commented on trino/trino_test.go:
```diff
@@ -765,6 +765,106 @@ func TestQueryColumns(t *testing.T) {
+  cast(current_time as time(6) with time zone) AS timeptz,
+  cast(current_time as timestamp(6)) AS tsp,
+  cast(current_time as timestamp(6) with time zone) AS tsptz`,
+		sql.Named("X-Trino-Client-Capabilities", "PARAMETRIC_DATETIME"))
```
Can this be enabled by default? This would simplify usage and reduce cognitive load - I don't remember how exactly this parameter behaves and would have to check Trino sources or other clients. I wouldn't expect users of this driver to know this.

Whatever we decide here, the README also should be updated.

hashhar replied:
In every client which supports this we want it to be the default behaviour - the Python client will start supporting this in next release and also defaults to always use the PARAMETRIC_DATETIME behaviour. It's a "breaking change" in some sense but IMO the benefits of always using parametric_datetime outweigh possible breakage.

JPMoresmau replied:
I have no problem enabling this since I set it in my own code, let me know if it's OK and if there should be a mechanism to not pass it. I stumbled upon this while investigating why my timestamps were truncated to milliseconds: https://github.com/trinodb/trino/issues/5068

hashhar replied:
In Python client we bit the bullet and enabled it by default with no option to fallback (since I can't think of any case where the alternative behaviour of precision always being 3 is useful).

I'd like to hear @losipiuk's opinion as well on this.

JPMoresmau replied:
Enabled by default, let me know.

JPMoresmau replied:
README updated to document the date/time support

nineinchnick commented on trino/trino_test.go:
```diff
@@ -765,6 +765,106 @@ func TestQueryColumns(t *testing.T) {
+	columns, err := rows.Columns()
+	require.NoError(t, err, "Failed reading result columns")
+
+	assert.Equal(t, 4, len(columns), "Expected 4 result column")
```
Is this necessary? Wouldn't `assert.Equal` called on a slice verify this?

JPMoresmau replied:
I just copied the test code from another test :-)

JPMoresmau replied:
We build another slice later on using that length, so I guess it's safer to verify it...

nineinchnick commented on trino/trino_test.go:
```diff
@@ -765,6 +765,106 @@ func TestQueryColumns(t *testing.T) {
+		{
+			"TIMESTAMP WITH TIME ZONE",
+			false,
+			0,
```
Shouldn't this be non-zero?

JPMoresmau replied:
Added support to capture timestamp and time column precision.

nineinchnick commented on trino/trino_test.go:
```diff
@@ -765,6 +765,106 @@ func TestQueryColumns(t *testing.T) {
+  cast(current_time as time(6)) AS timep,
+  cast(current_time as time(6) with time zone) AS timeptz,
+  cast(current_time as timestamp(6)) AS tsp,
+  cast(current_time as timestamp(6) with time zone) AS tsptz`,
```
What's the expected behavior for a high precision, like 12? If we can't store such values and they would be truncated, an error should be raised, or they should be returned as strings.

JPMoresmau replied:
We now error if the precision is greater than nanoseconds

nineinchnick commented on trino/trino_test.go:
```diff
@@ -1070,6 +1206,42 @@ func TestTypeConversion(t *testing.T) {
+		{
+			DataType:                   "timestamp with time zone",
+			RawType:                    "timestamp with time zone",
+			ResponseUnmarshalledSample: "2017-07-10 01:02:03.123456789 +03:00",
```
Can the Trino server send a geographical TZ, like `Europe/Warsaw`, not just an offset?

JPMoresmau replied:
Yes, I think so, I'll check this works.

JPMoresmau replied:
Added a couple of tests, this works.

nineinchnick commented on trino/trino.go:
```diff
@@ -1863,16 +1867,21 @@ func (s *NullSlice3Float64) Scan(value interface{}) error {
 }
 
 // Layout for time and timestamp WITH time zone.
+// Trino can support up to 12 digits sub second precision.
```
Can we really store picoseconds in a `time.Time`? [The docs](https://pkg.go.dev/time#Time) say:
> A Time represents an instant in time with nanosecond precision.

JPMoresmau replied:
So but we need to be able to parse the string containing 12 digits into Go. And yes I guess the value will get truncated because Go doesn't support better than nanos

nineinchnick replied:
The driver should not have to parse it just to throw away picoseconds. If it cannot handle picoseconds, it's reasonable to expect queries to be written in a way to avoid returning them (either round or truncate explicitly) and return an actionable error message explaining this.

JPMoresmau replied:
We now error on precision greater than nanoseconds. So reading a timestamp as `TIMESTAMP(9)` works but greater precision fails.

nineinchnick commented:
Looks like this would resolve #13

JPMoresmau commented on trino/serial.go:
```diff
@@ -35,6 +35,21 @@ func (e UnsupportedArgError) Error() string {
 type Numeric string
 
+// Date indicates we want a Date type in Trino from a Golang time.
+type Date time.Time
```
OK, so instead of just a type wrapper we make Date and Time more explicit structs with only the fields they need? And keep the time wrappers for timestamps?

nineinchnick replied:
How they store time internally would be an implementation detail. I guess it only matters if we want to support nano or picoseconds. I was thinking about how to avoid surprises when using them.

Allowing to use `time.Time` directly would be nice too.

JPMoresmau replied:
I have replaced the type wrappers with functions that explicitly request the various fields they contain, or make it explicit what they strip, let me know. I've also removed TimestampTz but added direct support for time.Time (since according to the doc time.Time is ALWAYS carrying a timezone, since nil location means UTC).

nineinchnick commented on trino/serial.go:
```diff
@@ -92,7 +107,19 @@ func Serial(v interface{}) (string, error) {
 		return "", UnsupportedArgError{"[]byte"}
 
+	case Date:
+		return "DATE " + time.Time(x).Format("'2006-01-02'"), nil
```
You might create a `Date` somewhere from a `time.Time` that you created earlier or received from another part of a program and not know it has a non-zero time value. Consider a function like this:
```
func getOrders(since time.Time) []Order {
    rows, err := conn.Query(`SELECT * FROM orders WHERE created_at > ?`, trino.Date(since))
    // ...
}
```
This is error-prone, and the driver should not allow this. If we'd require the callers to explicitly create the Date from parts of `time.Time`, like `trino.Date(since.Year(), since.Month(), since.Day())` it's obvious to the caller that the time part is being ignored.

JPMoresmau replied:
Done

JPMoresmau commented on trino/serial_test.go:
```diff
@@ -113,6 +116,31 @@ func TestSerial(t *testing.T) {
 		},
+		{
+			name:           "date",
+			value:          Date(time.Date(2017, 7, 10, 0, 0, 0, 0, time.Local)),
```
Since the way to construct these has been changed, there is no hidden truncation

nineinchnick replied:
I can't resolve comments in PRs in this repository, feel free to mark them as resolved yourself (I think you should be able to).

nineinchnick commented on trino/serial.go:
```diff
@@ -35,6 +35,55 @@ func (e UnsupportedArgError) Error() string {
+	day   int
+}
+
+func Date(year int, month time.Month, day int) trinoDate {
```
All exported functions should have comments, maybe something like:
```suggestion
// Date creates a representation of a Trino Date type.
func Date(year int, month time.Month, day int) trinoDate {
```

nineinchnick commented on trino/serial.go:
```diff
@@ -35,6 +35,55 @@ func (e UnsupportedArgError) Error() string {
+	nanosecond int
+}
+
+func Time(hour int,
```
Same here. Also, I wonder if it would be too much to describe what to do when a higher precision is needed.
```suggestion
// Time creates a representation of a Trino Time type. To represent time with precision higher than nanoseconds, pass the value as a string and use a cast in the query.
func Time(hour int,
```

nineinchnick commented on trino/serial.go:
```diff
@@ -92,9 +141,18 @@ func Serial(v interface{}) (string, error) {
-		return "", UnsupportedArgError{"time.Time"}
+		return "TIMESTAMP " + time.Time(x).Format("'2006-01-02 15:04:05.999999999 Z07:00'"), nil
+
+		// TODO - support durations
```
The TODO seems unnecessary, there's an informative error being returned.

nineinchnick commented on trino/trino.go:
```diff
@@ -733,10 +737,11 @@ func (st *driverStmt) QueryContext(ctx context.Context, args []driver.NamedValue
 	query := st.query
-	var hs http.Header
+	hs := make(http.Header)
+	// Ensure we have full date precision support.
```
I checked [the docs](https://trino.io/docs/current/develop/client-protocol.html) to see if we could copy the description but it doesn't mention this header so we can't reference it here. How about this:
```suggestion
	// Ensure the server returns timestamps preserving their precision, without truncating them to timestamp(3).
```

nineinchnick commented on trino/serial_test.go:
```diff
@@ -14,9 +14,16 @@
+)
 
 func TestSerial(t *testing.T) {
+	loc, err := time.LoadLocation("Europe/Paris")
```
```suggestion
	paris, err := time.LoadLocation("Europe/Paris")
```

nineinchnick commented on trino/trino_test.go:
```diff
@@ -765,6 +765,123 @@ func TestQueryColumns(t *testing.T) {
+
+	if shouldError {
+		assert.False(t, rows.Next())
+		require.Error(t, rows.Err())
```
It would be more readable as two separate tests and you should test for a specific error message, not just the fact that any error is present. If you'd do that, test cases in one of the tests would be simpler as it would not have to check for precision.

JPMoresmau replied:
We can still perfectly get the precision for the columns query even though retrieving the actual rows errors. Also the error message is very technical, and not too helpful because it shows the parsing error for the last layout we tried. Should have a more legible error like "value XXX cannot be parsed as a date or time"? Or collect the various errors into one?

nineinchnick replied:
> We can still perfectly get the precision for the columns query even though retrieving the actual rows errors.

This is an even better reason to split these up (or just have a separate test for errors).

> Should have a more legible error like "value XXX cannot be parsed as a date or time"

That would be ideal, but if you think it's out of the scope of this PR we can address it later. Is it still possible to test for a specific error message, even if it's a complex one? I mean can we keep the tests simple?

JPMoresmau replied:
Yes, I've changed the test that errors to use a time literal so we can test for a consistent error message since the order of layouts is deterministic. Maybe a better error message could be done in a separate PR, yes.

nineinchnick reviewed (commented):
Just a few nitpicks, almost there!

nineinchnick commented on README.md:
```diff
@@ -215,7 +220,8 @@ SELECT * FROM table WHERE col_double = cast(? AS DOUBLE) OR col_timestamp = CAST
 When reading response rows, the driver supports most Trino data types, except:
-* time and timestamps with precision - all time types are returned as `time.Time`
+* time and timestamps with precision - all time types are returned as `time.Time`.
+  All precisions up to nanoseconds (`TIMESTAMP(9)` or `TIME(9)`) are supported.  
```
That can be a bit out of the scope of this PR, since what I requested should go into the description of `trino.NullTime`, in the [Response rows](https://github.com/trinodb/trino-go-client#response-rows) section, but there's not much there now. This should not block this PR from getting merged.

JPMoresmau commented:
@nineinchnick I think I've merged your suggestions and addressed your comments, let me know. I suppose i can squash the commits once you're fine with my latest changes...

nineinchnick commented on trino/serial.go:
```diff
@@ -35,6 +35,57 @@ func (e UnsupportedArgError) Error() string {
+// Timestamp indicates we want a TimeStamp type WITHOUT a time zone in Trino from a Golang time.
+type trinoTimestamp time.Time
+
+func StripTimezone(time time.Time) trinoTimestamp {
+	return trinoTimestamp(time)
```
You've applied my suggestion to change this in the README but not here.
```suggestion
// Timestamp creates a representation of a Trino Timestamp(9) type.
func Timestamp(year int,
    month time.Month,
    day int,
    hour int,
	minute int,
	second int,
	nanosecond int) trinoTimestamp {
	return trinoTimestamp(time.Date(day, month, year, hour, minute, second, nanosecond, time.UTC)
```

nineinchnick commented on trino/serial.go:
```diff
@@ -35,6 +35,57 @@ func (e UnsupportedArgError) Error() string {
+	return trinoDate{year, month, day}
+}
+
+// Time indicates we want a Time type in Trino from a Golang time.
```
```suggestion
// trinoTime represents a Time type in Trino.
```

nineinchnick commented on trino/serial.go:
```diff
@@ -35,6 +35,57 @@ func (e UnsupportedArgError) Error() string {
+	return trinoTime{hour, minute, second, nanosecond}
+}
+
+// TimeTz indicates we want a Time With Timezone type in Trino from a Golang time.
```
```suggestion
// trinoTimeTz represents a Time(9) With Timezone type in Trino.
```

nineinchnick commented on trino/serial.go:
```diff
@@ -35,6 +35,57 @@ func (e UnsupportedArgError) Error() string {
+// TimeTz indicates we want a Time With Timezone type in Trino from a Golang time.
+type trinoTimeTz time.Time
+
+func TimeTz(hour int,
```
```suggestion
// TimeTz creates a representation of a Trino Time(9) With Timezone type.
func TimeTz(hour int,
```

nineinchnick commented on trino/trino_test.go:
```diff
@@ -765,6 +765,321 @@ func TestQueryColumns(t *testing.T) {
+
+}
+
+func testPrecisionDateTime(t *testing.T, precision int64, shouldError bool) {
```
This is now unused, remove it.

nineinchnick reviewed (commented):
Thanks for bearing with me. This is the last round from me. LGTM % these nitpicks.

I'm not a maintainer, so we need @hashhar and/or @losipiuk to do the review now. You could squash commits to make reviewing easier for them.

losipiuk commented on README.md:
```diff
@@ -215,7 +220,11 @@ SELECT * FROM table WHERE col_double = cast(? AS DOUBLE) OR col_timestamp = CAST
 When reading response rows, the driver supports most Trino data types, except:
-* time and timestamps with precision - all time types are returned as `time.Time`
+* time and timestamps with precision - all time types are returned as `time.Time`.
+  All precisions up to nanoseconds (`TIMESTAMP(9)` or `TIME(9)`) are supported (since
```
why exactly are we constraining ourselve to nanosecond precision? Would it be possible to support up to `TIME*(12)`?

nineinchnick replied:
Because Go's `time.Time` only supports nanoseconds. To support picoseconds, we'd have to define a custom data type. I'm not sure how much work it would be to make it compatible with `time.Time`.

nineinchnick replied:
Recent work in the Python client in https://github.com/trinodb/trino-python-client/pull/300 is also limited but to microseconds.

losipiuk commented on trino/serial.go:
```diff
@@ -35,6 +35,65 @@ func (e UnsupportedArgError) Error() string {
+	return trinoTime{hour, minute, second, nanosecond}
+}
+
+// trinoTimeTz represents a Time(9) With Timezone type in Trino.
```
nit:
```
trinoTimeTz represents a Time With Timezone type in Trino with precision up to 9
```

losipiuk commented on trino/serial.go:
```diff
@@ -35,6 +35,65 @@ func (e UnsupportedArgError) Error() string {
+// trinoTimeTz represents a Time(9) With Timezone type in Trino.
+type trinoTimeTz time.Time
+
+// TimeTz creates a representation of a Trino Time(9) With Timezone type.
```
nit: like above

losipiuk commented on trino/serial.go:
```diff
@@ -35,6 +35,65 @@ func (e UnsupportedArgError) Error() string {
+// Timestamp indicates we want a TimeStamp type WITHOUT a time zone in Trino from a Golang time.
+type trinoTimestamp time.Time
+
+// Timestamp creates a representation of a Trino Timestamp(9) type.
```
those comments are not really useful? @nineinchnick do we have check which enforces those?

nineinchnick replied:
We don't currently have a linter in the CI that would check for this. The built-in `go vet` doesn't, we'd have to use `staticcheck` (recommended after `go lint` was deprecated) or `golangci-lint` (which is huge but popular).

Having comments for all exported functions is idiomatic in Go and it would show up in online docs like here: https://pkg.go.dev/github.com/trinodb/trino-go-client@v0.308.0/trino

What would it take to make them more useful? Should it explain why `time.Time` is not used instead?

losipiuk replied:
> What would it take to make them more useful

Not sure. Presonally I would just drop it. But if we want to follow Go guildelines let's go with it :)

JPMoresmau commented:
Sorry, is anything more needed from me to get that PR approved?
